recipes/perl/rbac_roles/rbac_role_templates.pl

```diff
@@ -0,0 +1,793 @@
+        'template'    => \%security_role_operation_map,
+}
+);
+
+# Parse command line options
+my ($help, $man, $debug);
+GetOptions(
+    'help|h!'          => \$help,
+    'man|m!'           => \$man,
+    'debug|d!',        => \$debug,
+    'list_templates'   => \$list_templates,
+    'hostname|host=s'  => \$hostname,
+    'username|user=s'  => \$username,
+    'password|pass=s'  => \$password,
+    'domain_type=s'    => \$domain_type,
+    'domain_name=s'    => \$domain_name,
+    'token=s'          => \$token,
+    'create_templates' => \$create_all,
+    'template=s'       => \$template,
+    'name=s'           => \$name,
+    'description=s'    => \$description
+) or pod2usage(2);
+
+if ($help) {
+    pod2usage(1);
+    exit;
+}
+
+if ($man) {
+    pod2usage(-verbose => 2);
+    exit;
+}
+
+if ($list_templates) {
+    print_templates();
+    exit;
+}
+
+unless ($create_all or $template) {
+    pod2usage(1);
+    exit 1;
+}
+
+if ($create_all and $template) {
+    print "Cannot specify both --create_templates and --template for role creation.\n";
+    pod2usage(1);
+    exit 1;
+}
+
+if ($create_all and ($name or $description)) {
+    print "--name and --description are only valid with the --template option.\n";
+    pod2usage(1);
+    exit 1;
+}
+
+if ($template) {
+    my $found = 0;
+    foreach my $role (keys %role_template_data) {
+        if ($template =~ /$role/) {
+            $found = 1;
+        }
+    }
+
+    unless ($found) {
+        print "Invalid --template specified [$template]\n";
+        print "Valid templates are:\n";
+        print_templates();
+        exit 1;
+    }
+}
+
+if ($name) {
+    $role_template_data{$template}{name} = $name;
+    print "Role name provided: " . $role_template_data{$template}{name} . "\n";
+}
+
+if ($description) {
+    $role_template_data{$template}{description} = $description;
+    print "Role description provided: " . $role_template_data{$template}{description} . "\n";
+}
+
+# Initialize HTTP client
+my $http = LWP::UserAgent->new(
+    ssl_opts => {verify_hostname => 0, verify_peer => 0},
+    protocols_allowed => ['https']
+);
+my $response;
+
+# Ping server to verify connection and get API version
+my $ping_url = "https://$hostname/netbackup/ping";
+$response = $http->get($ping_url, 'Accept' => 'text/plain');
+unless ($response->is_success) {
+    if ($debug) {
+        print Dumper($response);
+    }
+    die "Failed to ping server!\n" ;
+}
+
+my $api_version = $response->header('x-netbackup-api-version');
+if ($api_version < 4) {  # NetBackup 8.3
+    die "Script should not be run on NetBackup releases before 8.3!\n";
+}
+
+# Set media type with discovered API version
+my $media_type = "application/vnd.netbackup+json;version=$api_version";
+
+# Log in to get a token if needed
+unless ($token) {
+
+    unless ($password) {
+        ReadMode('noecho');
+        print "Password:";
+        $password = ReadLine(0);
```

Comment:
Two issues here.

1. Don't forget to `ReadMode 'restore'` before exiting the program. Best place would be immediately after this line.
2. I suspect care will be required to restore the input mode in case of Ctrl+C. An unexpected password prompt seems like it would be a pretty common place for a user to decide to cancel, and doing so shouldn't leave the terminal unusable. Off the top of my head, I don't know how to address that, not whether it's worth addressing anyway.